### ISSUE TITLE
Cost functions in MLP with minibatch

### ISSUE THREAD

**Issue by albertmolinermrf** [haifengl/smile#525]
## Expected behaviour

The cost function in neural networks is typically defined as the expected value (i.e. the average) of the per-example cost function over the training set.

## Actual behaviour

The cost function is the sum of the per-example cost function over the training set.

In online learning the two functions above are equal. When a minibatch is used then they differ by a factor equal to the size of the minibatch. Obviously, this does not affect the optimization algorithm itself, as the minimum of any function is located at the same point as the minimum of the function multiplied by a constant. However, the effective learning rate is the nominal learning rate multiplied by the size of the minibatch, which can significantly affect results if hyperparameters have been set before the selection of the batch size.

**Comment by haifengl** [comment]
There is no golden rule how to set the learning rate.

**Comment by haifengl** [comment]
It is worth of mentioning that the average operation will make the algorithm slower. Since we are of pure Java and have disadvantages on numerical computation, it is important not to introduce any extra unnecessary computations. It is the user's responsibility to set the proper learning rate.

**Comment by haifengl** [comment]
in 2.5.3, the gradient is averaged over the minibatch.
